File: src/LTAdmin/Services/Business/EvaluationService.cs
using LTAdmin.Core;
using LTAdmin.Data;
using LTAdmin.Models.Dto;
using LTAdmin.Models.Entities;
using LTAdmin.Repositories;
using LTAdmin.Services.Common;
using LTAdmin.Services.Logging;
using LTAdmin.Services.Referentiel;
using LTAdmin.Services.Validation;

namespace LTAdmin.Services.Business;

/// <summary>
/// Gestion des périodes d'évaluation et des évaluations (devoirs, interros).
/// Règles : barème et poids strictement positifs, période clôturée figée.
/// </summary>
public sealed class EvaluationService : ServiceBase
{
    private readonly EvaluationRepository _evaluations;
    private readonly StaffRepository _personnel;

    public EvaluationService(AccessDatabase database, AppLogger logger, JournalService journal, ParametreService parametres)
        : base(database, logger, journal, parametres)
    {
        _evaluations = new EvaluationRepository(database);
        _personnel = new StaffRepository(database);
    }

    public List<PeriodeEval> ListPeriodes(int? idAnnee = null)
    {
        try { return _evaluations.ListPeriodes(idAnnee); }
        catch (Exception ex)
        {
            Logger.Error("Liste des périodes d’évaluation", ex);
            return new List<PeriodeEval>();
        }
    }

    public PeriodeEval? GetPeriode(int idPeriode)
    {
        try { return _evaluations.GetPeriode(idPeriode); }
        catch (Exception ex)
        {
            Logger.Error("Lecture d’une période", ex);
            return null;
        }
    }

    public Result<int> CreatePeriode(PeriodeEval periode, string codeUtr)
    {
        var validation = EvaluationValidator.ValidatePeriode(periode);
        if (!validation.IsValid) return Invalid<int>(validation);
        try
        {
            var id = _evaluations.InsertPeriode(periode);
            Journal.LogCreation(codeUtr, Tables.PeriodeEval, id, periode.Libelle);
            return Result<int>.Ok(id, "Période d’évaluation créée.");
        }
        catch (Exception ex) { return Failure<int>("Création d’une période", ex); }
    }

    public Result UpdatePeriode(PeriodeEval periode, string codeUtr)
    {
        if (!periode.IdPeriode.HasValue)
            return Result.Fail("Période introuvable.", ErrorCodes.NotFound);
        var validation = EvaluationValidator.ValidatePeriode(periode);
        if (!validation.IsValid) return Invalid(validation);
        try
        {
            var updated = _evaluations.UpdatePeriode(periode);
            if (updated == 0)
                return Result.Fail("Période introuvable.", ErrorCodes.NotFound);
            Journal.LogModification(codeUtr, Tables.PeriodeEval, periode.IdPeriode, periode.Libelle);
            return Result.Ok("Période enregistrée.");
        }
        catch (Exception ex) { return Failure("Modification d’une période", ex); }
    }

    /// <summary>Clôture (ou rouvre) une période : les notes figées ne sont plus modifiables.</summary>
    public Result SetPeriodeCloturee(int idPeriode, bool cloturee, string codeUtr)
    {
        try
        {
            var periode = _evaluations.GetPeriode(idPeriode);
            if (periode is null)
                return Result.Fail("Période introuvable.", ErrorCodes.NotFound);
            _evaluations.SetPeriodeCloturee(idPeriode, cloturee);
            Journal.LogOperation(codeUtr, cloturee ? "CLOTURE_PERIODE" : "REOUVERTURE_PERIODE",
                Tables.PeriodeEval, idPeriode, periode.Libelle);
            return Result.Ok(cloturee ? "Période clôturée : les notes sont figées." : "Période rouverte.");
        }
        catch (Exception ex) { return Failure("Clôture d’une période", ex); }
    }

    public List<EvaluationDetail> ListEvaluations(int? idPeriode = null, int? idClasse = null, int? idProg = null)
    {
        try { return _evaluations.ListEvaluations(idPeriode, idClasse, idProg); }
        catch (Exception ex)
        {
            Logger.Error("Liste des évaluations", ex);
            return new List<EvaluationDetail>();
        }
    }

    public EvaluationDetail? GetEvaluationDetail(int idEvaluation)
    {
        try { return _evaluations.GetEvaluationDetail(idEvaluation); }
        catch (Exception ex)
        {
            Logger.Error("Lecture d’une évaluation", ex);
            return null;
        }
    }

    public Result<int> CreateEvaluation(Evaluation evaluation, string codeUtr)
    {
        if (!evaluation.Bareme.HasValue || evaluation.Bareme.Value <= 0)
            evaluation.Bareme = Parametres.BaremeDefaut;
        if (!evaluation.Poids.HasValue || evaluation.Poids.Value <= 0)
            evaluation.Poids = 1d;
        var validation = EvaluationValidator.ValidateEvaluation(evaluation);
        if (!validation.IsValid) return Invalid<int>(validation);

        try
        {
            var periode = _evaluations.GetPeriode(evaluation.IdPeriode!.Value);
            if (periode is null)
                return Result<int>.Fail("Période introuvable.", ErrorCodes.NotFound);
            if (periode.Cloturee == true)
                return Result<int>.Fail("Période clôturée : rouvrez-la pour ajouter une évaluation.", ErrorCodes.BusinessRule);
            var programme = _personnel.GetProgramme(evaluation.IdProg!.Value);
            if (programme is null)
                return Result<int>.Fail("Programme (classe × matière) introuvable.", ErrorCodes.NotFound);

            var id = _evaluations.InsertEvaluation(evaluation);
            Journal.LogCreation(codeUtr, Tables.Evaluation, id, evaluation.Intitule);
            return Result<int>.Ok(id, "Évaluation créée.");
        }
        catch (Exception ex) { return Failure<int>("Création d’une évaluation", ex); }
    }

    public Result UpdateEvaluation(Evaluation evaluation, string codeUtr)
    {
        if (!evaluation.IdEvaluation.HasValue)
            return Result.Fail("Évaluation introuvable.", ErrorCodes.NotFound);
        var validation = EvaluationValidator.ValidateEvaluation(evaluation);
        if (!validation.IsValid) return Invalid(validation);

        try
        {
            var periode = _evaluations.GetPeriode(evaluation.IdPeriode!.Value);
            if (periode?.Cloturee == true)
                return Result.Fail("Période clôturée : rouvrez-la pour modifier cette évaluation.", ErrorCodes.BusinessRule);
            var updated = _evaluations.UpdateEvaluation(evaluation);
            if (updated == 0)
                return Result.Fail("Évaluation introuvable.", ErrorCodes.NotFound);
            Journal.LogModification(codeUtr, Tables.Evaluation, evaluation.IdEvaluation, evaluation.Intitule);
            return Result.Ok("Évaluation enregistrée.");
        }
        catch (Exception ex) { return Failure("Modification d’une évaluation", ex); }
    }

    public Result SetPubliee(int idEvaluation, bool publiee, string codeUtr)
    {
        try
        {
            var existing = _evaluations.GetEvaluation(idEvaluation);
            if (existing is null)
                return Result.Fail("Évaluation introuvable.", ErrorCodes.NotFound);
            _evaluations.SetEvaluationPubliee(idEvaluation, publiee);
            Journal.LogOperation(codeUtr, publiee ? "PUBLICATION_NOTES" : "DEPUBLICATION_NOTES",
                Tables.Evaluation, idEvaluation, existing.Intitule);
            return Result.Ok(publiee ? "Notes publiées." : "Publication retirée.");
        }
        catch (Exception ex) { return Failure("Publication d’une évaluation", ex); }
    }

    public Result DeleteEvaluation(int idEvaluation, string codeUtr)
    {
        try
        {
            var existing = _evaluations.GetEvaluation(idEvaluation);
            if (existing is null)
                return Result.Fail("Évaluation introuvable.", ErrorCodes.NotFound);
            _evaluations.DeleteEvaluation(idEvaluation);
            Journal.LogSuppression(codeUtr, Tables.Evaluation, idEvaluation, existing.Intitule);
            return Result.Ok("Évaluation supprimée.");
        }
        catch (Exception ex)
        {
            if (OleDbExceptionHelper.IsForeignKeyViolation(ex))
                return Result.Fail("Suppression impossible : des notes sont déjà saisies pour cette évaluation.",
                    ErrorCodes.ForeignKeyViolation);
            return Failure("Suppression d’une évaluation", ex);
        }
    }
}

File: ltadmin/services/business/evaluation_service.py
"""Gestion des périodes d'évaluation et des évaluations (devoirs, interros).

Règles : barème et poids strictement positifs, période clôturée figée.
"""

from __future__ import annotations

from typing import List, Optional

from ltadmin.core.result import Result, ResultValue
from ltadmin.data import error_helper
from ltadmin.data.access_database import AccessDatabase
from ltadmin.data.schema import Tables
from ltadmin.models.dto import EvaluationDetail
from ltadmin.models.entities import Evaluation, PeriodeEval
from ltadmin.repositories.evaluation_repository import EvaluationRepository
from ltadmin.repositories.staff_repository import StaffRepository
from ltadmin.services.common import ServiceBase
from ltadmin.services.logging.app_logger import AppLogger
from ltadmin.services.logging.journal_service import JournalService
from ltadmin.services.referentiel.parametre_service import ParametreService
from ltadmin.services.validation.validation import EvaluationValidator


class EvaluationService(ServiceBase):

    def __init__(self, database: AccessDatabase, logger: AppLogger,
                 journal: JournalService, parametres: ParametreService):
        super().__init__(database, logger, journal, parametres)
        self._evaluations = EvaluationRepository(database)
        self._personnel = StaffRepository(database)

    # ----- Périodes -----

    def list_periodes(self, id_annee: Optional[int] = None) -> List[PeriodeEval]:
        try:
            return self._evaluations.list_periodes(id_annee)
        except Exception as ex:
            self.logger.error("Liste des périodes d’évaluation", ex)
            return []

    def get_periode(self, id_periode: int) -> Optional[PeriodeEval]:
        try:
            return self._evaluations.get_periode(id_periode)
        except Exception as ex:
            self.logger.error("Lecture d’une période", ex)
            return None

    def create_periode(self, periode: PeriodeEval, code_utr: str) -> ResultValue[int]:
        if periode.id_annee is None:
            # Rattachement automatique à l'année scolaire active.
            from ltadmin.repositories.admin_repository import AdminRepository
            annee = AdminRepository(self.db).get_annee_active()
            if annee is None:
                return ResultValue.fail(
                    "Aucune année scolaire active : activez-en une dans "
                    "l’administration.", "GESTION")
            periode.id_annee = annee.id_annee
        validation = EvaluationValidator.validate_periode(periode)
        if not validation.is_valid:
            return self.invalid_value(validation)
        try:
            id_periode = self._evaluations.insert_periode(periode)
            self.journal.log_creation(code_utr, Tables.PERIODE_EVAL, id_periode, periode.libelle)
            return ResultValue.ok(id_periode, "Période d’évaluation créée.")
        except Exception as ex:
            return self.failure_value("Création d’une période", ex)

    def update_periode(self, periode: PeriodeEval, code_utr: str) -> Result:
        if periode.id_periode is None:
            return Result.fail("Période introuvable.", "INTROUVABLE")
        validation = EvaluationValidator.validate_periode(periode)
        if not validation.is_valid:
            return self.invalid(validation)
        try:
            updated = self._evaluations.update_periode(periode)
            if updated == 0:
                return Result.fail("Période introuvable.", "INTROUVABLE")
            self.journal.log_modification(
                code_utr, Tables.PERIODE_EVAL, periode.id_periode, periode.libelle)
            return Result.ok("Période enregistrée.")
        except Exception as ex:
            return self.failure("Modification d’une période", ex)

    def set_periode_cloturee(self, id_periode: int, cloturee: bool, code_utr: str) -> Result:
        """Clôture (ou rouvre) une période : les notes figées ne sont plus modifiables."""
        try:
            periode = self._evaluations.get_periode(id_periode)
            if periode is None:
                return Result.fail("Période introuvable.", "INTROUVABLE")
            self._evaluations.set_periode_cloturee(id_periode, cloturee)
            self.journal.log_operation(
                code_utr,
                "CLOTURE_PERIODE" if cloturee else "REOUVERTURE_PERIODE",
                Tables.PERIODE_EVAL, id_periode, periode.libelle)
            return Result.ok(
                "Période clôturée : les notes sont figées." if cloturee
                else "Période rouverte.")
        except Exception as ex:
            return self.failure("Clôture d’une période", ex)

    # ----- Évaluations -----

    def list_evaluations(self, id_periode: Optional[int] = None,
                         id_classe: Optional[int] = None,
                         id_prog: Optional[int] = None) -> List[EvaluationDetail]:
        try:
            return self._evaluations.list_evaluations(id_periode, id_classe, id_prog)
        except Exception as ex:
            self.logger.error("Liste des évaluations", ex)
            return []

    def get_evaluation_detail(self, id_evaluation: int) -> Optional[EvaluationDetail]:
        try:
            return self._evaluations.get_evaluation_detail(id_evaluation)
        except Exception as ex:
            self.logger.error("Lecture d’une évaluation", ex)
            return None

    def create_evaluation(self, evaluation: Evaluation, code_utr: str) -> ResultValue[int]:
        if evaluation.bareme is None or evaluation.bareme <= 0:
            evaluation.bareme = self.parametres.bareme_defaut
        if evaluation.poids is None or evaluation.poids <= 0:
            evaluation.poids = 1.0
        validation = EvaluationValidator.validate_evaluation(evaluation)
        if not validation.is_valid:
            return self.invalid_value(validation)
        try:
            periode = self._evaluations.get_periode(evaluation.id_periode)
            if periode is None:
                return ResultValue.fail("Période introuvable.", "INTROUVABLE")
            if periode.cloturee is True:
                return ResultValue.fail(
                    "Période clôturée : rouvrez-la pour ajouter une évaluation.", "GESTION")
            programme = self._personnel.get_programme(evaluation.id_prog)
            if programme is None:
                return ResultValue.fail(
                    "Programme (classe × matière) introuvable.", "INTROUVABLE")
            evaluation.id_evaluation = self._evaluations.insert_evaluation(evaluation)
            self.journal.log_creation(
                code_utr, Tables.EVALUATION, evaluation.id_evaluation,
                evaluation.intitule)
            return ResultValue.ok(evaluation.id_evaluation, "Évaluation créée.")
        except Exception as ex:
            return self.failure_value("Création d’une évaluation", ex)

    def update_evaluation(self, evaluation: Evaluation, code_utr: str) -> Result:
        if evaluation.id_evaluation is None:
            return Result.fail("Évaluation introuvable.", "INTROUVABLE")
        validation = EvaluationValidator.validate_evaluation(evaluation)
        if not validation.is_valid:
            return self.invalid(validation)
        try:
            periode = self._evaluations.get_periode(evaluation.id_periode)
            if periode is not None and periode.cloturee is True:
                return Result.fail(
                    "Période clôturée : rouvrez-la pour modifier cette évaluation.", "GESTION")
            updated = self._evaluations.update_evaluation(evaluation)
            if updated == 0:
                return Result.fail("Évaluation introuvable.", "INTROUVABLE")
            self.journal.log_modification(
                code_utr, Tables.EVALUATION, evaluation.id_evaluation, evaluation.intitule)
            return Result.ok("Évaluation enregistrée.")
        except Exception as ex:
            return self.failure("Modification d’une évaluation", ex)

    def set_publiee(self, id_evaluation: int, publiee: bool, code_utr: str) -> Result:
        try:
            existing = self._evaluations.get_evaluation(id_evaluation)
            if existing is None:
                return Result.fail("Évaluation introuvable.", "INTROUVABLE")
            self._evaluations.set_evaluation_publiee(id_evaluation, publiee)
            self.journal.log_operation(
                code_utr,
                "PUBLICATION_NOTES" if publiee else "DEPUBLICATION_NOTES",
                Tables.EVALUATION, id_evaluation, existing.intitule)
            return Result.ok("Notes publiées." if publiee else "Publication retirée.")
        except Exception as ex:
            return self.failure("Publication d’une évaluation", ex)

    def delete_evaluation(self, id_evaluation: int, code_utr: str) -> Result:
        try:
            existing = self._evaluations.get_evaluation(id_evaluation)
            if existing is None:
                return Result.fail("Évaluation introuvable.", "INTROUVABLE")
            self._evaluations.delete_evaluation(id_evaluation)
            self.journal.log_suppression(
                code_utr, Tables.EVALUATION, id_evaluation, existing.intitule)
            return Result.ok("Évaluation supprimée.")
        except Exception as ex:
            if error_helper.is_foreign_key_violation(ex):
                return Result.fail(
                    "Suppression impossible : des notes sont déjà saisies pour cette "
                    "évaluation.", "LIAISON")
            return self.failure("Suppression d’une évaluation", ex)
